apis/placement/v1beta1/clusterresourceplacement_types.go

```diff
@@ -207,6 +219,55 @@ type TopologySpreadConstraint struct {
 	WhenUnsatisfiable UnsatisfiableConstraintAction `json:"whenUnsatisfiable,omitempty"`
 }
 
+// RolloutStrategy describes how to update existing placed resource either onto a different cluster
+// or in-place upgrade to the latest resource.
+type RolloutStrategy struct {
+	// Type of rollout. The only supported type is "RollingUpdate". Default is "RollingUpdate".
+	// +optional
+	Type RolloutStrategyType `json:"type,omitempty"`
+
+	// Rolling update config params. Present only if RolloutStrategyType = RollingUpdate.
+	// +optional
+	RollingUpdate *RollingUpdateConfig `json:"rollingUpdate,omitempty"`
+}
+
+// +enum
+type RolloutStrategyType string
+
+const (
+	// RollingUpdateRolloutStrategyType replace the old placed resource using rolling update
+	// i.e. gradually create the new one while replace the old ones.
+	RollingUpdateRolloutStrategyType RolloutStrategyType = "RollingUpdate"
+)
+
+// RollingUpdateConfig contains the config to control the desired behavior of rolling update.
+type RollingUpdateConfig struct {
+	// The maximum number of clusters that can be unavailable during the rolling update
+	// comparing to the desired number of clusters.
+	// Absolute number is calculated from percentage by applying to the new target number of clusters.
+	// We consider a resource not available when we either remove it from a cluster or in-place
+	// upgrade the resources content on the same cluster.
+	// This can not be 0 if MaxSurge is 0.
+	// Defaults to 25%.
+	// +optional
+	MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty"`
+
+	// The maximum number of clusters that can be scheduled above the desired number of clusters.
+	// Absolute number is calculated from percentage by applying to the new target number of clusters.
+	// This does not apply to the case that we do in-place upgrade of resources on the same cluster.
+	// This can not be 0 if MaxUnavailable is 0.
+	// Defaults to 25%.
+	// +optional
+	MaxSurge *intstr.IntOrString `json:"maxSurge,omitempty"`
+
+	// WaitBetweenRollingPeriodSeconds is used to config the time to wait between rolling out phases
+	// A resource placement is considered available if `WaitBetweenRollingPeriodSeconds` seconds
+	// has passed after the resources are applied to the target cluster successfully.
+	// Default is 60.
+	// +optional
+	WaitBetweenRollingPeriodSeconds *int `json:"waitBetweenRollingPeriodSeconds,omitempty"`
```

Comment:
i'm not sure if i understand this field & MaxUnavailable correctly.
For example, there are 5 old replicas and MaxUnavailable = 3. rollout controller updates 3 replicas. After WaitBetweenRollingPeriodSeconds, the resources fails to apply and we treat them as unavailable = 3. What will rollout controller do in this case?

will it keep reconciling until one of resource placement succeeds to unblock other replica update?

Reply:
let's assume it's resource only rollout. The rollout will stop after this as there are already 3 unavailable.